src/v/cloud_storage/remote.cc

```diff
@@ -1401,8 +1401,10 @@ ss::future<remote::list_result> remote::list_objects(
 
             list_bucket_result.prefix = list_result.prefix;
 
-            // Continue to list the remaining items
-            if (items_remaining) {
+            // Continue to list the remaining items, while respecting max_keys.
```

Comment:
If I read this correctly, if we get a continuation token but the caller didn't specify max_keys then we stop at whatever length of the result the provider provides. For AWS that might be 1000, for another that might be 100 (Cloudflare R2) doesn't even document it.

I believe we should flip this around. If the caller of remote doesn't specify max_keys, then likely it wasn't ready to handle truncated results so we should keep iterating as we were doing before. It doesn't seem like that code worked correctly but I believe the intention was good and we should keep it.

We can give a special behavior to max_keys = 0 here as a signal from the caller that they are ready to handle truncation. Then, we don't forward max_keys to the backing APIs/clients and we stop on the first result/whatever the provider defined limit is.

If you audit the callers of `remote::list_objects` you'll notice that none of them check the `list_bucket_result::is_truncated` property...

Reply:
Oop, yeah there is a bug here with the comparison of max_keys when it is `std::nullopt`. Current test coverage is insufficient, I will have to add a few tests.

Reply:
I have implemented this behaviour for truncation when `max_keys = 0`. Details in the PR description and my commit messages.